File: kumi/edge/template/kumi_tools/typescript/kumiSetup.ts
/**
 * Kumi Edge — TypeScript tool registration
 *
 * Register your functions with `agent.register()` and call `initKumi()`
 * from your app entry point.
 *
 *     import { initKumi } from "./kumi_tools/typescript/kumiSetup";
 *     initKumi();
 *
 * Quick test: ``npx tsx kumiSetup.ts`` (from kumi_tools/typescript/)
 *
 * Requires: npm install (from the kumi_tools/typescript/kumi_sdk directory)
 */

import path from "node:path";
import { pathToFileURL } from "node:url";

import { KumiAgent } from "./kumi_sdk/src";

// ── Connection (edit here, or set in .env) ──

const KUMI_CONNECTION_CODE = "kumi-lan_..."; // paste from `kumi --server`, or kumi_... for relay
const KUMI_EDGE_NAME = "My Node App";

export function initKumi(): KumiAgent {
  const agent = new KumiAgent({
    connectionCode: KUMI_CONNECTION_CODE,
    edgeName: KUMI_EDGE_NAME,
  });

  // ── Register tools: name + description + parameters + handler ──

  // agent.register({
  //   name: "jump",
  //   description: "Make the character jump",
  //   parameters: [
  //     { name: "height", type: "number", description: "Jump height in meters" },
  //   ],
  //   handler: async (args) => {
  //     const height = args.number("height") ?? 1.0;
  //     return `Jumped ${height} meters`;
  //   },
  // });

  // Dangerous tools: user confirms in the Kumi web UI or `kumi --chat` (not on device):
  // agent.register({
  //   name: "delete_all",
  //   description: "Delete all data",
  //   requireConfirmation: true,
  //   handler: async () => "Deleted everything",
  // });

  agent.runInBackground();
  return agent;
}

/** True when this file is executed directly (e.g. `npx tsx kumiSetup.ts`). */
function isDirectRun(): boolean {
  if (typeof process === "undefined" || !process.argv[1]) {
    return false;
  }
  return import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
}

if (isDirectRun()) {
  initKumi();
  console.log("Kumi edge running (kumiSetup as entry). Press Ctrl+C to stop.");
}

File: kumi/edge/template/kumi_tools/python/kumi_setup.py
"""
Kumi Edge — Python tool registration

Import your own functions and register them with a description.
Parameter types are auto-extracted from type hints, and parameter
descriptions from the docstring Args section.

Usage — embed in your app::

    from kumi_tools.python.kumi_setup import init_kumi

    init_kumi()
    # Your program continues to run as usual

Quick test — run this file only (no separate main.py)::

    python -m kumi_tools.python.kumi_setup

    # or from ``kumi_tools/python/``:
    python kumi_setup.py

Requires: pip install websockets

Multi-tenant server (``kumi-enterprise``): set ``KUMI_ACCESS_TOKEN`` to your user
``kumi_...`` token if your deployment requires authenticated Edge registration.
"""

try:
    from .kumi_sdk import KumiAgent
except ImportError:
    # ``python kumi_setup.py`` from ``kumi_tools/python/`` (not as a package)
    from kumi_sdk import KumiAgent

# ── Import your tool functions ──
# from my_app.actions import jump, run


def init_kumi():
    agent = KumiAgent(
        # connection_code="kumi-lan_...",  # or set KUMI_CONNECTION_CODE in .env
        # edge_name="My Device",              # or set EDGE_NAME in .env
    )

    # ── Register tools: func + description ──
    # The description tells the AI when and how to use the tool.
    # Tool name and parameter types are auto-extracted from the function.
    #
    # agent.register(jump, "Make the character jump")
    # agent.register(run, "Make the character run at a given speed")
    #
    # Dangerous tools: user confirms in the Kumi web UI or `kumi --chat` (not on device):
    # agent.register(delete_all, "Delete all data", require_confirmation=True)
    #
    # High-value tools that should bypass dynamic routing every turn:
    # agent.register(get_status, "Read current app status", always_include=True)
    #
    # Read-only tools can opt in to proactive messaging context:
    # agent.register(get_status, "Read current app status", allow_proactive=True, proactive_context=True)
    #
    # Tool confirmation choices (Tools page / chat "always allow") are saved next to your
    # .env as .kumi_tool_confirmation.json (override with KUMI_TOOL_CONFIRMATION_PATH).

    agent.run_in_background()
    return agent


if __name__ == "__main__":
    import sys
    import threading

    init_kumi()
    print("Kumi edge running (setup as __main__). Press Ctrl+C to stop.", file=sys.stderr)
    try:
        threading.Event().wait()
    except KeyboardInterrupt:
        pass
